pkg/distributor/otel.go

```diff
@@ -277,30 +278,33 @@ func otlpHandler(
 }
 
 // toOtlpGRPCHTTPStatus is utilized by the OTLP endpoint.
-// According to the OTLP specifications (https://opentelemetry.io/docs/specs/otlp/#failures-1), unlike Prometheus, the OTLP client only retries on HTTP status codes 429, 502, 503, and 504.
 func toOtlpGRPCHTTPStatus(pushErr error) (codes.Code, int) {
-	if !errors.Is(pushErr, context.DeadlineExceeded) {
-		var distributorErr Error
-		if errors.As(pushErr, &distributorErr) {
-			switch distributorErr.Cause() {
-			case mimirpb.BAD_DATA:
-				return codes.InvalidArgument, http.StatusBadRequest
-			case mimirpb.INGESTION_RATE_LIMITED, mimirpb.REQUEST_RATE_LIMITED:
-				return codes.ResourceExhausted, http.StatusTooManyRequests
-			case mimirpb.REPLICAS_DID_NOT_MATCH:
-				return codes.OK, http.StatusAccepted
-			case mimirpb.TOO_MANY_CLUSTERS:
-				return codes.InvalidArgument, http.StatusBadRequest
-			case mimirpb.TSDB_UNAVAILABLE:
-				return codes.Unavailable, http.StatusServiceUnavailable
-			case mimirpb.CIRCUIT_BREAKER_OPEN:
-				return codes.Unavailable, http.StatusServiceUnavailable
-			case mimirpb.METHOD_NOT_ALLOWED:
-				return codes.Unimplemented, http.StatusNotImplemented
-			}
+	var distributorErr Error
+	if errors.Is(pushErr, context.DeadlineExceeded) || !errors.As(pushErr, &distributorErr) {
+		return codes.Internal, http.StatusServiceUnavailable
+	}
+
+	grpcStatusCode := errorCauseToGRPCStatusCode(distributorErr.Cause(), false)
+	httpStatusCode := errorCauseToHTTPStatusCode(distributorErr.Cause(), false)
+	otlpHTTPStatusCode := httpRetryableToOtlpRetryable(httpStatusCode)
+	return grpcStatusCode, otlpHTTPStatusCode
+}
+
+// httpRetryableToOtlpRetryable maps non-retryable 5xx HTTP status codes according
+// to the OTLP specifications (https://opentelemetry.io/docs/specs/otlp/#failures-1)
+// to http.StatusServiceUnavailable. In case of a non-retryable HTTP status code,
+// httpRetryableToOtlpRetryable returns the HTTP status code itself.
+// Unlike Prometheus, which retries 429 and all 5xx HTTP status codes,
+// the OTLP client only retries on HTTP status codes 429, 502, 503, and 504.
+func httpRetryableToOtlpRetryable(httpStatusCode int) int {
```

Comment:
[Style nit]
```suggestion
// httpRetryableToOTLPRetryable maps non-retryable 5xx HTTP status codes according
// to the OTLP specifications (https://opentelemetry.io/docs/specs/otlp/#failures-1)
// to http.StatusServiceUnavailable. In case of a non-retryable HTTP status code,
// httpRetryableToOTLPRetryable returns the HTTP status code itself.
// Unlike Prometheus, which retries 429 and all 5xx HTTP status codes,
// the OTLP client only retries on HTTP status codes 429, 502, 503, and 504.
func httpRetryableToOTLPRetryable(httpStatusCode int) int {
```